File: mdc_lib/python/examples/极简控制/mdc_lite.py
# -*- coding: utf-8 -*-
"""
mdc_lite.py — 极简调用库（只管调用 / send-only，独立实现）
=============================================================

**独立实现**：自带 CRC8、组帧与命令打包，**不依赖任何外部库**（不 import mdc_lib）。
只用到 Python 标准库 `struct`。

**核心思想：上位机调参、下位机执行。** 你只需要告诉下位机"发什么控制量"，
本模块把"要发送的帧字节"打包好交给你，由你自行 `ser.write()` 发送。不解析回包。

**只涉及 3 条命令**（帧格式 [0xAA][CMD][LEN][DATA][CRC8]，CRC 范围=CMD+LEN+DATA，多项式 0x07 初值 0）：
    * 0x31 MOTOR_CTRL    四通道控制目标（下位机执行）
    * 0x40 SUBSCRIBE     开启状态周期上报（收速度的前提）
    * 0x41 UNSUBSCRIBE   关闭状态上报（善后）

**用法（串口由你实现）：**

    import serial, mdc_lite
    ser = serial.Serial("COM5", 2000000)
    ser.write(mdc_lite.subscribe(50))             # 先订阅（如需回读转速）
    ser.write(mdc_lite.ctrl(100, -200, 0, 300))   # 四通道目标值
    ser.write(mdc_lite.ctrl(0, 0, 0, 0))          # 归零
    ser.write(mdc_lite.unsubscribe())             # 退出前取消订阅（可选）

**如需读回转速，见 `mdc_lite_ctrl`（调用+回调接收，同为独立实现）。**
"""

import struct

# ---- 常量 ----
MD_SYNC = 0xAA            # 帧同步字
MD_MAX_DATA = 250         # DATA 段最大长度
MD_CRC8_POLY = 0x07       # CRC8 多项式（初值 0）
MD_CMD_MOTOR_CTRL = 0x31  # 四通道控制
MD_CMD_SUBSCRIBE = 0x40   # 订阅状态上报
MD_CMD_UNSUBSCRIBE = 0x41 # 取消订阅

__all__ = ["ctrl", "stop", "subscribe", "unsubscribe"]


def _crc8(data):
    """CRC8：多项式 0x07，初值 0，按位计算。校验向量 crc8([0x01,0x00])==0x15。"""
    crc = 0
    for b in data:
        crc ^= b
        for _ in range(8):
            if crc & 0x80:
                crc = ((crc << 1) ^ MD_CRC8_POLY) & 0xFF
            else:
                crc = (crc << 1) & 0xFF
    return crc


def _build_frame(cmd, data=b""):
    """组帧 [0xAA][CMD][LEN][DATA...][CRC8]，CRC=CMD+LEN+DATA。"""
    body = bytes((cmd, len(data))) + data
    return bytes((MD_SYNC,)) + body + bytes((_crc8(body),))


def ctrl(m0, m1, m2, m3):
    """0x31 MOTOR_CTRL：四通道控制目标（int32 LE），返回完整帧 bytes。

    例：``ctrl(100, -200, 0, 300)`` 的 DATA 段 == 64 00 00 00 38 FF FF FF 00 00 00 00 2C 01 00 00。
    """
    return _build_frame(MD_CMD_MOTOR_CTRL, struct.pack("<4i", m0, m1, m2, m3))


def stop():
    """便捷：四通道全零控制帧（急停/退出前发送，0x31 DATA=16B 全零）。"""
    return ctrl(0, 0, 0, 0)


def subscribe(interval_ms):
    """0x40 SUBSCRIBE：开启状态周期上报，返回完整帧 bytes。

    :param interval_ms: 上报周期 ms，建议 >=20（固件钳位）。
    """
    return _build_frame(MD_CMD_SUBSCRIBE, struct.pack("<H", interval_ms))


def unsubscribe():
    """0x41 UNSUBSCRIBE：关闭状态上报，返回完整帧 bytes。"""
    return _build_frame(MD_CMD_UNSUBSCRIBE)

File: mdc_lib/esp32/arduino/mdc_lite.cpp
/**
 * @file    mdc_lite.cpp
 * @brief   Motor Driver Controller 极简调用库 —— 只管调用（send-only，独立实现）
 *
 * 规范依据：../LITE.md（mdc_lite 极简 LITE API）
 *
 * **独立实现**：本文件自带 CRC8 与全部组帧逻辑，不复用本目录 mdc_lib.cpp
 * （不调用 md_bin_motor_ctrl / md_bin_subscribe / md_bin_unsubscribe）。
 * 协议字节布局（LE）、CRC8-ATM（0x07 初值 0）与帧格式
 * [0xAA][CMD][LEN][DATA...][CRC8] 与 mdc_lib / API.md 完全一致，仅实现
 * "极简 4 条命令"的发送侧。写入 out 缓冲并返回字节数；cap 不足返回 0。
 *
 * 本文件不 include 任何 Arduino / ESP 头文件（纯 C++，C++11 兼容）。
 *
 * 四平台（esp32 / rp2040 / avr / esp8266）此文件内容完全一致。
 */

#include "mdc_lite.h"

/* ==================== 内部工具：小端手动拼装（16 位 int 平台如 AVR 上行为一致） ==================== */

static inline void md_lite_put_u32(uint8_t* p, uint32_t v)
{
    p[0] = (uint8_t)(v & 0xFF);
    p[1] = (uint8_t)((v >> 8) & 0xFF);
    p[2] = (uint8_t)((v >> 16) & 0xFF);
    p[3] = (uint8_t)((v >> 24) & 0xFF);
}

/* ==================== 底层：CRC8 ==================== */

uint8_t md_lite_crc8(const uint8_t* data, uint16_t len)
{
    uint8_t c = 0;

    if (data == nullptr)
        return 0;

    for (uint16_t i = 0; i < len; i++) {
        c ^= data[i];
        for (uint8_t b = 0; b < 8; b++) {
            if (c & 0x80)
                c = (uint8_t)((c << 1) ^ MD_CRC8_POLY);
            else
                c = (uint8_t)(c << 1);
        }
    }
    return c;
}

/* ==================== 内部工具：组帧 ==================== */

/* 组帧：[0xAA][CMD][LEN][DATA...][CRC8]，CRC 范围 = CMD+LEN+DATA（不含 SYNC）。
 * 返回写入字节数；data_len>MD_MAX_DATA 或 cap 不足/参数非法返回 0。 */
static inline uint16_t md_lite_build_frame(uint8_t cmd, const uint8_t* data,
                                           uint16_t data_len, uint8_t* out, uint16_t cap)
{
    uint16_t total;

    if (out == nullptr || data_len > (uint16_t)MD_MAX_DATA)
        return 0;
    total = (uint16_t)(data_len + 4);   /* SYNC + CMD + LEN + DATA + CRC */
    if (cap < total)
        return 0;

    out[0] = MD_SYNC;
    out[1] = cmd;
    out[2] = (uint8_t)data_len;
    if (data_len > 0) {
        if (data == nullptr)
            return 0;
        for (uint16_t i = 0; i < data_len; i++)
            out[3 + i] = data[i];
    }
    out[total - 1] = md_lite_crc8(out + 1, (uint16_t)(data_len + 2));   /* CMD+LEN+DATA */
    return total;
}

/* ==================== 发送侧（send-only） ==================== */

uint16_t md_lite_ctrl(int32_t m0, int32_t m1, int32_t m2, int32_t m3,
                      uint8_t* out, uint16_t cap)
{
    uint8_t data[16];
    int32_t t[4];

    if (out == nullptr || cap < 20)
        return 0;

    t[0] = m0; t[1] = m1; t[2] = m2; t[3] = m3;
    for (uint16_t i = 0; i < 4; i++)
        md_lite_put_u32(data + i * 4, (uint32_t)t[i]);

    /* 0x31 四通道控制帧。校验向量：ctrl(100,-200,0,300)
     * DATA == 64 00 00 00 38 FF FF FF 00 00 00 00 2C 01 00 00（整帧 AA 31 10 ... E7） */
    return md_lite_build_frame(MD_CMD_MOTOR_CTRL, data, (uint16_t)sizeof(data), out, cap);
}

uint16_t md_lite_stop(uint8_t* out, uint16_t cap)
{
    /* 便捷：四通道全零控制帧（急停/退出前发送，DATA=16B 全零） */
    return md_lite_ctrl(0, 0, 0, 0, out, cap);
}

uint16_t md_lite_subscribe(uint16_t interval_ms, uint8_t* out, uint16_t cap)
{
    uint8_t data[2];

    if (out == nullptr || cap < 6)
        return 0;

    data[0] = (uint8_t)(interval_ms & 0xFF);
    data[1] = (uint8_t)((interval_ms >> 8) & 0xFF);

    /* 0x40 订阅状态上报。校验向量：subscribe(50) 帧 == AA 40 02 32 00 9E */
    return md_lite_build_frame(MD_CMD_SUBSCRIBE, data, (uint16_t)sizeof(data), out, cap);
}

uint16_t md_lite_unsubscribe(uint8_t* out, uint16_t cap)
{
    /* 0x41 取消订阅。校验向量：unsubscribe() 帧 == AA 41 00 4E */
    return md_lite_build_frame(MD_CMD_UNSUBSCRIBE, nullptr, 0, out, cap);
}
